### ISSUE TITLE
Allow to use custom C++ objectives from python API classes `CatBoostRegressor`, `CatBoostClassifier` and `CatBoostRanker`

### ISSUE THREAD

**Issue by andrey-khropov** [catboost/catboost#2936]
Right now it does not work because `loss_function="UserPerObjMetric"` does not pass `_check_is_compatible_loss` checks.

**Comment by isdg** [comment]
Hi, @andrey-khropov! This is an interesting issue and I wanted to clarify some things if possible :)

- Is this internal or either open-source aimed issue? 
- Probably not correct question but how you see custom cpp objective usage in `_fit` of python API classes? Passing `loss_function="UserPerObjMetric"` is clear but even if it would pass `_check_is_compatible_loss` what can be properly done with abstract class `TUserDefinedPerObjectError ` https://github.com/catboost/catboost/blob/2cd145ed130f51b8cc39a4dc22bbd0321789ee03/catboost/private/libs/algo/tensor_search_helpers.cpp#L296-L299

Thanks in advance

**Comment by andrey-khropov** [comment]
> * Is this internal or either open-source aimed issue?

What do you mean? This repository is about open source CatBoost.

> what can be properly done with abstract class TUserDefinedPerObjectError 

It is [not abstact](https://github.com/catboost/catboost/blob/689c4f81ee5fb1082e91dda49007e981164e88c0/catboost/private/libs/algo_helpers/error_functions.h#L1199), otherwise the code couldn't be compiled.

See [the tutorial](https://github.com/catboost/tutorials/blob/master/custom_loss/custom_metric_tutorial.md#adding-custom-per-object-objective-function-tutorial).

But this approach while fast if implemented properly (code works at C++ speed) is not very flexible. Only a single function with the hardcoded name `UserPerObjMetric` is supported and you have to modify the source code and rebuild CatBoost.

If you use CatBoost python package a more flexible approach is to implement custom objectives and metrics in python code (and accelerate using `numba`). See [this tutorial](https://github.com/catboost/tutorials/blob/master/custom_loss/custom_loss_and_metric_tutorial.ipynb). Using this approach you won't have to re-build CatBoost itself and can have as many custom functions with different names as you need.

**Comment by isdg** [comment]
Thanks for your clarifications! Did not notice `final`. 

> What do you mean? This repository is about open source CatBoost.

I was referring to the [tutorial](https://github.com/catboost/tutorials/blob/master/custom_loss/custom_metric_tutorial.md#adding-custom-per-object-objective-function-tutorial). Since using a cpp custom loss requires rebuilding Catboost, I assumed it might be more for internal purposes, as it doesn’t seem very convenient for end users.

Just to confirm my understanding: the goal of the issue is to enable the following behaviour (assuming the end user has a rebuilt CatBoost), correct? And this would apply only to `loss_function`, not to `eval_metric`. Or is the goal more about increasing flexibility for cpp losses in general (even if there is already `numba` optimisation)?

```python
model = CatBoostRegressor(
    loss_function="UserPerObjMetric",
    # eval_metric="UserPerObjMetric".
)
model.fit(...)
preds = model.predict(...)
``` 

*CatBoostRegressor as an example*

**Comment by andrey-khropov** [comment]
> Just to confirm my understanding: the goal of the issue is to enable the following behaviour (assuming the end user has a rebuilt CatBoost), correct? 
> 
> model = CatBoostRegressor(
>     loss_function="UserPerObjMetric",
>     # eval_metric="UserPerObjMetric".
> )
> model.fit(...)
> preds = model.predict(...)
> 
> _CatBoostRegressor as an example_

Yes. Right now it works only with [`CatBoost` class](https://catboost.ai/docs/en/concepts/python-reference_catboost)

> And this would apply only to `loss_function`, not to `eval_metric`. 

This issue is about `loss_function`. `eval_metric` is not checked in `_check_is_compatible_loss`.

**Comment by isdg** [comment]
Can you assign this to me please? I think I can handle it.

**Comment by andrey-khropov** [comment]
> Can you assign this to me please? I think I can handle it.

Done. Thank you.
